In [2]:
s="안녕"
print(s, type(s))
s='Hello'
print(s)
s="""Hello
World"""
print(s)
s='''Hello
World'''
print(s) 
s='Hello\'s World'
print(s)
s="Hello's World"
print(s)
s="Hello\nWorld"
print(s)
s="Hello\tWorld"
print(s)

안녕 <class 'str'>
Hello
Hello
World
Hello
World
Hello's World
Hello's World
Hello
World
Hello	World


In [9]:
i=10
s=str(i)
print(s, type(s))
# s="i"+i  # 오류
s="i :"+str(i)
print(s)

10 <class 'str'>
i :10


In [10]:
i=10.5
s="i : {i}".format(i=i)
print(s)
s=f"i : {i}"
print(s)

i : 10.5
i : 10.5


In [ ]:
s="안녕"
s2="안녕"
print(s==s2)
print(s is s2)


True
False


In [35]:
s="Hello, World!"
print(s.upper())
print(s.lower())
print(s.replace("World", "Python"))
print(s.split(","))
print(s.strip())
print(s.find("World"))
# print(s.index("a")) # 오류 발생 : 반드시 해당 문자가 존재한다”는 전제로 동작하기 때문에 없으면 **예외(ValueError)
print(s.find("a")) # -1 반환
print(s.startswith("Hello"))
print(s.endswith("!"))
print(s.count("o")) 

HELLO, WORLD!
hello, world!
Hello, Python!
['Hello', ' World!']
Hello, World!
7
-1
True
True
2


In [60]:
s="Hello, World!"

print("H" in s)
print(s.find("H"))
print(s.index("H"))
print(s.count("o"))

True
0
0
2


In [ ]:
import re

s = "hello python"
m = re.match("he", s)
print(m)   # <re.Match ...>
m=re.search("py", s)
print(m)   # <re.Match ...>
m=re.findall(r"\d+", "age 20 height 180")
print(m)   # ['20', '180']
m=re.fullmatch(r"\d{3}-\d{4}", "010-1234")
print(m)   


<re.Match object; span=(0, 2), match='he'>
<re.Match object; span=(6, 8), match='py'>
['20', '180']
<re.Match object; span=(0, 8), match='010-1234'>
010-1234


In [29]:
s="abcdefg"
print(len(s))
print(s[0])
print(s[2])
print(s[-1])
print(s[-3])
print(s[2:5])
print(s[:4])
print(s[3:])
print(s[:])
print(s[::2])
print(s[1::2])          

7
a
c
g
e
cde
abcd
defg
abcdefg
aceg
bdf


In [ ]:
# 바이트 문자열, 바이트 리터럴
s=b"Hello, World!"
print(s)
print(s[0], type(s[0]))
print(s[0],s[1],s[2],s[3],s[4])

b'Hello, World!'
72 <class 'int'>
72 101 108 108 111


In [61]:
# s=b"안녕" # 오류 발생 : 바이트 리터럴은 ASCII 문자만 가능
s="안녕".encode("utf-8")
print(s, type(s))
print(s[0], type(s[0]))
print(s[0],s[1],s[2],s[3],s[4],s[5])
s2=s.decode("utf-8")
print(s2, type(s2))

b'\xec\x95\x88\xeb\x85\x95' <class 'bytes'>
236 <class 'int'>
236 149 136 235 133 149
안녕 <class 'str'>


## UTF-8에서 “문자 하나의 시작”을 알 수 있는 방법

UTF-8은 가변 길이의 인코딩으로 문자 하나가 1~4바이트까지 차지한다.

UTF-8의 규칙

각 바이트 앞부분 비트 패턴을 보면 “이 바이트가 새 문자 시작인지” 알 수 있음.

바이트 타입	비트 패턴	의미
ASCII	0xxxxxxx	문자 1바이트짜리, 시작
첫 바이트(2바이트 문자)	110xxxxx	새 문자 시작
첫 바이트(3바이트 문자)	1110xxxx	새 문자 시작
첫 바이트(4바이트 문자)	11110xxx	새 문자 시작
이어지는 바이트	10xxxxxx	문자 내부 바이트 (시작 아님)

236 → 0xEC → 11101100
235 → 0xEB → 11101011